src/procedure/nodes/select.cpp

```diff
@@ -264,25 +260,31 @@ bool SelectProcedureNode::execute(const ProcedureContext &procedureContext)
             // Check distance from reference site (if defined)
             if (distanceRef)
             {
-                r = procedureContext.configuration()->box()->minimumDistance(site.origin(), distanceRef->origin());
-                if (!inclusiveDistanceRange_.contains(r))
+                if (!inclusiveDistanceRange_.contains(
+                        procedureContext.configuration()->box()->minimumDistance(site.origin(), distanceRef->get().origin())))
                     continue;
             }
 
-            // All OK, so add site
-            sites_.push_back(&site);
+            // All OK, so add site and index in its stack (1...N numbering)
+            sites_.emplace_back(site, siteIndex, n + 1);
         }
     }
 
-    // Set first site index and increase selections counter
-    currentSiteIndex_ = (sites_.empty() ? -1 : 0);
+    // Increase selections counter
     ++nSelections_;
 
     // If a ForEach branch has been defined, process it for each of our sites in turn. Otherwise, we're done.
+    currentSite_ = std::nullopt;
     if (!forEachBranch_.empty())
     {
-        for (currentSiteIndex_ = 0; currentSiteIndex_ < sites_.size(); ++currentSiteIndex_)
+        auto index = 1;
+        for (const auto &siteInfo : sites_)
         {
+            currentSite_ = std::get<0>(siteInfo);
+            siteIndexParameter_->setValue(std::get<1>(siteInfo));
+            stackIndexParameter_->setValue(std::get<1>(siteInfo));
```

Comment:
Should this be `2` instead?
```suggestion
            stackIndexParameter_->setValue(std::get<2>(siteInfo));
```

Reply:
Oops!